alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs

```diff
@@ -623,3 +644,139 @@ instance
+        Just (PlutusScript bytes) -> pure $ Just bytes
+      pure (sp, sh, msb)
+
+    txinfo = txInfo ei sysS utxo tx
+
+    exBudgetToExUnits :: P.ExBudget -> Maybe ExUnits
+    exBudgetToExUnits (P.ExBudget (P.ExCPU cpu) (P.ExMemory memory)) =
+      ExUnits <$> convertCost cpu <*> convertCost memory
+
+    convertCost :: P.CostingInteger -> Maybe Word64
+    convertCost ci =
+      if (i <= toInteger (maxBound :: Word64)) && (i >= 0)
+        then Just $ fromIntegral i
+        else Nothing
+      where
+        i = toInteger ci
+
+    findAndCount ::
+      ( Alonzo.ScriptPurpose (Crypto era),
+        ScriptHash (Crypto era),
+        Maybe ShortByteString
+      ) ->
+      ( ScriptHash (Crypto era),
+        Either (ScriptFailure (Crypto era)) (RdmrPtr, ExUnits)
+      )
+    findAndCount (sp, sh, mScriptBytes) = (,) sh $ do
+      r <-
+        note (InvalidScriptPurpose sp) $
+          strictMaybeToMaybe $ Alonzo.rdptr @era txb sp
+      (rdmr, _) <- note (MissingRedeemer r) $ Map.lookup r rs
+      script <- note (MissingScript sh) $ mScriptBytes
+      args <- case sp of
+        (Alonzo.Spending txin) -> do
+          txOut <- note (UnknownTxIn txin) $ Map.lookup txin (unUTxO utxo)
+          let TxOut _ _ mdh = txOut
+          dh <- note (InvalidTxIn txin) $ strictMaybeToMaybe mdh
+          dat <- note (MissingDatum dh) $ Map.lookup dh dats
+          pure [dat, rdmr, valContext txinfo sp]
+        _ -> pure [rdmr, valContext txinfo sp]
+      let pArgs = map getPlutusData args
+
+      exUnits <- case snd $ P.evaluateScriptCounting P.Quiet costModel script pArgs of
+        Left e -> Left $ ValidationFailed e
+        Right exBudget -> note (IncompatibleBudget exBudget) $ exBudgetToExUnits exBudget
+
+      pure (r, exUnits)
+
+evaluateTransactionBalance ::
+  forall era.
+  ( Era era,
+    HasField "certs" (Core.TxBody era) (StrictSeq (DCert (Crypto era))),
+    HasField "inputs" (Core.TxBody era) (Set (TxIn (Crypto era))),
+    HasField "mint" (Core.TxBody era) (Core.Value era),
+    HasField "wdrls" (Core.TxBody era) (Wdrl (Crypto era)),
+    HasField "_keyDeposit" (Core.PParams era) Coin,
+    HasField "_poolDeposit" (Core.PParams era) Coin
+  ) =>
+  Core.PParams era ->
+  UTxO era ->
+  Map (KeyHash 'StakePool (Crypto era)) (PoolParams (Crypto era)) ->
+  Core.TxBody era ->
+  Core.Value era
+evaluateTransactionBalance pp u stakepools txb = consumed pp u txb Val.<-> Shelley.produced @era pp stakepools txb
+
+evaluateTransactionFee ::
+  forall era tx.
+  ( DSIGNAlgorithm (Crypto era),
+    HasField "_minfeeA" (Core.PParams era) Natural,
+    HasField "_minfeeB" (Core.PParams era) Natural,
+    HasField "_prices" (Core.PParams era) Prices,
+    HasField "totExunits" tx ExUnits,
+    HasField "txsize" tx Integer
+  ) =>
+  Core.PParams era ->
+  tx ->
+  Word ->
+  Coin
+evaluateTransactionFee pp tx numKeyWits =
+  minfee @era pp tx Val.<+> (Coin . fromIntegral $ numKeyWits * keyWitSize)
```

Comment:
I just realized that this is both incorrect, and still up for discussion. If we did use `numKeyWits`, we need `a * (txsize tx + numKeyWits * keyWitSize) + b`. But there are other options for adding witnesses later (as we talked about in slack). Multi-sig / Timelocks is what makes the other options hard.